File: bench/benches/primes.go
package main

func isPrime(n int64) bool {
	for d := int64(2); d*d <= n; d++ {
		if n%d == 0 {
			return false
		}
	}
	return true
}

func count(lo, hi int64) int64 {
	var c int64
	for n := lo; n < hi; n++ {
		if isPrime(n) {
			c++
		}
	}
	return c
}

func main() {
	bench("primes", func() int64 { return count(2, 30000) })
}

File: bench/benches/primes.hs
{-# LANGUAGE BangPatterns #-}
import Bench

isPrime :: Int -> Bool
isPrime n = go 2
  where
    go !d
      | d * d > n      = True
      | n `mod` d == 0 = False
      | otherwise      = go (d + 1)

count :: Int -> Int -> Integer
count lo hi = go lo 0
  where
    go !n !c
      | n >= hi   = c
      | isPrime n = go (n + 1) (c + 1)
      | otherwise = go (n + 1) c

main :: IO ()
main = bench "primes" (\u -> count (with u 2) (with u 30000))
